controllers/setupcontroller/setup_controller.go

```diff
@@ -0,0 +1,127 @@
+package setupcontroller
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	dscv1 "github.com/opendatahub-io/opendatahub-operator/v2/apis/datasciencecluster/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/cluster"
+	odhClient "github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/client"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/upgrade"
+)
+
+type SetupControllerReconciler struct {
+	*odhClient.Client
+}
+
+const (
+	finalizerName = "datasciencecluster.opendatahub.io/finalizer"
+)
+
+func (r *SetupControllerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
The entire logic of deleting the `DSC` can probably be moved to the  `upgrade.OperatorUninstall`, so this method can be simplified to something like:

```go
if !upgrade.HasDeleteConfigMap(ctx, r.Client) {
    return ctrl.Result{}, nil
}

if err := upgrade.OperatorUninstall(ctx, r.Client, cluster.GetRelease().Name); err != nil {
    return ctrl.Result{}, fmt.Errorf("error while operator uninstall: %w", err)
}

return ctrl.Result{}, nil
```

Reply:
Yeah, I was thinking the same, but I am just trying to get a successful test on the current changes after which I'll make this refactor